kv/kvserver: TestRangeInfoReturned failed

kv/kvserver.TestRangeInfoReturned [failed](https://tanzanite.cluster.engflow.com/invocations/default/e264ea70-67cb-43c4-b753-03da30923a39?testReportRun=1&testReportShard=18&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [7571549d8772a087ab4577d1e770e02582ba5877](https://github.com/cockroachdb/cockroach/commits/7571549d8772a087ab4577d1e770e02582ba5877):

Fatal error:

```
panic: mutex is not write locked [recovered]
	panic: mutex is not write locked [recovered]
	panic: mutex is not write locked [recovered]
	panic: mutex is not write locked [recovered]
	panic: mutex is not write locked

```
Stack: 

```
goroutine 23578 [running]:
testing.tRunner.func1.2({0xa440ec0, 0xe8285e0})
	GOROOT/src/testing/testing.go:1631 +0x3f7
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x6b6
panic({0xa440ec0?, 0xe8285e0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:137 +0x4fa
panic({0xa440ec0?, 0xe8285e0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc006e1c590?, {0xe8a1c48, 0x1464d940})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x74
panic({0xa440ec0?, 0xe8285e0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc00fa879a0, {0xe8a1c48, 0x1464d940})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:521 +0x28b
panic({0xa440ec0?, 0xe8285e0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/syncutil.(*Mutex).AssertHeld(...)
	github.com/cockroachdb/cockroach/pkg/util/syncutil/mutex_sync_race.go:61
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaForRACv2).RaftMuAssertHeld(0xc005cf5608)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/flow_control_replica_integration.go:467 +0x51
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2.(*processorImpl).OnDescChangedLocked(0xc008370a00, {0xe8a1c48?, 0x1464d940?}, 0xc0013be310, {0xc00807ba4a?})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/processor.go:580 +0x70
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).setDescLockedRaftMuLocked(0xc005cf5608, {0xe8a1c48, 0x1464d940}, 0xc0013be310)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_init.go:437 +0x11c5
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).setDescRaftMuLocked(0xc005cf5608, {0xe8a1c48, 0x1464d940}, 0xc0013be310)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_init.go:366 +0xbf
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*testContext).addBogusReplicaToRangeDesc(0xc004a6ad20, {0xe8a1c48, 0x1464d940})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_test.go:306 +0x7a6
github.com/cockroachdb/cockroach/pkg/kv/kvserver.TestRangeInfoReturned(0xc004fb89c0)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_test.go:13874 +0x2e5
testing.tRunner(0xc004fb89c0, 0xb6a51c0)
	GOROOT/src/testing/testing.go:1689 +0x21f
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1742 +0x826
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRangeInfoReturned
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRangeInfoReturned1712086330
    test_log_scope.go:81: use -show-logs to present logs inline
*
* ERROR: a panic has occurred!
* panic: mutex is not write locked
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:770
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:521
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:770
*   | github.com/cockroachdb/cockroach/pkg/util/syncutil.(*Mutex).AssertHeld
*   | 	github.com/cockroachdb/cockroach/pkg/util/syncutil/mutex_sync_race.go:61
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaForRACv2).RaftMuAssertHeld
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/flow_control_replica_integration.go:467
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2.(*processorImpl).OnDescChangedLocked
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/processor.go:580
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).setDescLockedRaftMuLocked
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_init.go:437
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).setDescRaftMuLocked
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_init.go:366
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*testContext).addBogusReplicaToRangeDesc
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_test.go:306
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.TestRangeInfoReturned
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_test.go:13874
*   | testing.tRunner
*   | 	GOROOT/src/testing/testing.go:1689
*   | runtime.goexit
*   | 	src/runtime/asm_amd64.s:1695
* Wraps: (2) panic: mutex is not write locked
* Error types: (1) *withstack.withStack (2) *errutil.leafError
*
    stopper.go:230: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: outputs.zip/logTestRangeInfoReturned1712086330
    stopper.go:230: panic: mutex is not write locked
--- FAIL: TestRangeInfoReturned (0.06s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=18</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangeInfoReturned.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/kv/kvserver/replica_test.go
+++ b/pkg/kv/kvserver/replica_test.go
@@ -303,8 +303,8 @@ func (tc *testContext) addBogusReplicaToRangeDesc(
 		return roachpb.ReplicaDescriptor{}, err
 	}
 
-	tc.repl.setDescRaftMuLocked(ctx, &newDesc)
 	tc.repl.raftMu.Lock()
+	tc.repl.setDescRaftMuLocked(ctx, &newDesc)
 	tc.repl.mu.RLock()
 	tc.repl.assertStateRaftMuLockedReplicaMuRLocked(ctx, tc.engine)
 	tc.repl.mu.RUnlock()

